rfd/0231-new-oracle-join-method.md

```diff
@@ -0,0 +1,309 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 231 - New Oracle Join Method
+
+## Required Approvers
+
+* Engineering: @rosstimothy && @strideynet
+* Security: @rob-picard-teleport
+* Product: @klizhentas
+
+## What
+
+This RFD proposes a new join method for OCI (Oracle Cloud Infrastructure)
+compute instances.
+
+## Why
+
+A major drawback of our
+[current Oracle join method](https://goteleport.com/docs/enroll-resources/agents/oracle/)
+is that it requires all joining instances to be added to a dynamic IAM group,
+and granting the following policy to the group:
+
+```
+Allow dynamic-group '<identity domain>'/'join-teleport' to inspect authentication in tenancy
+```
+
+This policy is used to sign requests to the following API:
+
+```
+https://auth.<region>.oraclecloud.com/v1/authentication/authenticateClient
+```
+
+This API is undocumented by Oracle, and when Oracle learned we were using it for
+this purpose and requiring this policy on all instances, they assisted us in
+finding another method using supported APIs that will have public documentation
+soon.
+
+## Details
+
+### UX
+
+The new design will strictly improve UX as compared to the existing Oracle join
+method.
+Instances will not be required to have membership in any IAM groups and no IAM
+policies will be required.
+This will decrease the configuration burden and eliminate security questions of
+why the policy is necessary.
+
+Configuration on the Teleport side will not change.
+The new design will continue to leverage existing Oracle join tokens that
+Teleport admins may have configured.
+No configuration options will need to change on the agent or bot.
+
+### Design
+
+The new method will leverage OCI instance identity certificates, which are made
+available to all OCI compute instances via the IMDS (Instance Metadata Service).
+The best docs available on this certificate are the FAQ entries
+[here](https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/callingservicesfrominstances.htm#faq).
+This certificate can be retrieved on any instance with the following `curl` command:
+
+```
+curl -H 'Authorization: Bearer Oracle' http://169.254.169.254/opc/v2/identity/cert.pem
```

Comment:
Is the inclusion of an Authorization header mandatory? I seem to recall that when I tested this last year, this wasn't necessary, so it'd be interesting if they've introduced it in the mean time.

Reply:
This is OCI's version of IMDSv2, they now strongly recommend disabling IMDSv1 which doesn't require the header

Maybe worth noting that OCI's IMDS v2 is much weaker than AWS IMDSv2, on OCI it's still a single GET request with a known constant header, whereas AWS requires a PUT to get a session token and including the session token in the authz header of subsequent requests